### ISSUE TITLE
Inconsistent PME energy when atoms cross the boundary

### ISSUE THREAD

**Issue by kcui7** [openmm/openmm#2500]
Hi all, 

I'm using OpenMM to study zeolitic imidazolate frameworks (ZIFs). I found that the PME energy is inconsistent when some atoms are at the same position, but in different periodic images. I attached an example here. In this example, the simulation box is cubic with dimension 16.960 angstrom. The only differences between two files `pos1.pdb` and `pos2.pdb` are the positions of two Zn atoms:

273c273
ATOM    272   Zn  ZN B   8      16.949   8.275   4.198  1.00  0.00          ZN  
ATOM    272   Zn  ZN B   8      -0.011   8.275   4.198  1.00  0.00          ZN  
277c277
ATOM    276   Zn  ZN B  12       0.030   3.988   0.005  1.00  0.00          ZN  
ATOM    276   Zn  ZN B  12       0.030   3.988  16.965  1.00  0.00          ZN 

with periodic boundary condition, these two PDB files are the same. However, when calculating the energy of the system using `PME`, OpenMM gives inconsistent results:
6568.66039984 kJ/mol for pos1.pdb and
6462.46465866 kJ/mol for pos2.pdb.

On the other hand, if I change the `NonbondedMethod` to `CutoffPeriodic`, the energies of the two system are the same:
3226.42034301 kJ/mol for pos1.pdb and
3226.42034301 kJ/mol for pos2.pdb 

I've also tried `Ewald`, the results are similar to those using `PME`.
The related files are attached below.
[pme_test.zip](https://github.com/openmm/openmm/files/3940820/pme_test.zip)

**Comment by kcui7** [comment]
I did another test. In this calculation I removed all Zn-N bonds, which may cross the boundaries. This time the energies of these two systems are the same. The related files are attached below. 

It seems that the issue is not from the PME, but from combining the PME with bonds cross the periodic boundary. 

[no_zn_im_bonds.zip](https://github.com/openmm/openmm/files/3941618/no_zn_im_bonds.zip)

**Comment by peastman** [comment]
Take a look at https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#periodic, particularly the discussion of bonded forces, and see if that explains what's going on.  By default periodic boundary conditions aren't applied to bonded forces, though you can enable that if you need it by calling `setUsesPeriodicBoundaryConditions()`.

**Comment by kcui7** [comment]
Hi Peter, 

I've tested the `setUsesPeriodicBoundaryConditions()` and I don't think the problem is from the bond terms. In my first example, the FF contains only charge parameters so the energy difference is purely from the Coulomb interaction. 

I've looked at the link you sent me. I think I have to break a molecule between two different periodic copies because I'm simulating a crystal. And the `make_molecules_whole()` function won't work for me.

**Comment by peastman** [comment]
I just tested the latest version of your script on two different computers, and with all four platforms.  In every case I got identical energies for the two PDB files.

```
$ python simrun.py pos1.pdb 
-7655.997523746367 kJ/mol
$ python simrun.py pos2.pdb 
-7655.997584781523 kJ/mol
```

What version of OpenMM are you using?  Also what platform, GPU, and OS?

**Comment by kcui7** [comment]
I do get the same answer for the last case, however, that script is just a testing script and can not be used for production simulations. 
The first example is more practical (with Zn-N bonds) and in that example I got different results.

**Comment by kcui7** [comment]
Let's rephrase my problem.  Let's first forget the ZIF stuffs and examples and look at this simplify case. 
[1Im_in_box.zip](https://github.com/openmm/openmm/files/3942089/1Im_in_box.zip)

Assume we have one imidazole molecule in a periodic box, and we have two systems. The first system `pos1.pdb`, the molecule is broken between two different periodic copies, (which we actually shouldn't do this) and for the second system `pos2.pdb`, the molecule is whole and unbroken. I used `setUsesPeriodicBoundaryConditions(True)` for both cases during the calculation. 

1. When calculating the energy, I get `-79.0439416679 kJ/mol` for the first system and `-47.6460243744 kJ/mol` for the second system. 
2. After commenting out the `NonbondedForce`, I get the same energy `7.48632390664 kJ/mol` for both systems. This suggests that the energy difference is from the `NonbondedForce`. 
3. With `NonbondedForce` but removing all bonds (or connections) in the molecule, I can also get the same results `895870.919751 kJ/mol` for both systems. This suggests that the problem is from the exclusions. 
4. However, I printed the exclusions of the `NonbondedForce` in both systems, and they are the same, which means the problem is not from the exclusions. 

I'm very confused about what's going on. My test 3 and 4 gives contradictory conclusions.

**Comment by peastman** [comment]
The difference is coming from the exclusions (1-2, 1-3, and 1-4 interactions).  Those are really bonded interactions, so periodic boundary conditions don't get applied to them.  And if you split a molecule between two periodic copies, that changes their energy.

**Comment by kcui7** [comment]
I see. The exclusions only apply to the nonbonded interaction within the  molecule.
Is there a solution to this? I have to split molecules when simulating a crystal.

**Comment by peastman** [comment]
Given the current feature set, I don't see any way to do that.  On the other hand, I don't think it would be that hard to add a `setExceptionsUsePeriodicBoundaryConditions()` option.  I'll mark this issue as an enhancement request to add that feature.

**Comment by kcui7** [comment]
Thank you so much!

**Comment by peastman** [comment]
I'm looking into implementing this.  I have some questions about what the desired behavior is.

Exceptions are normally computed without either cutoffs or periodic boundary conditions.  Nonbonded interactions use both cutoffs and periodic boundary conditions.  The cutoff ensures the forces are continuous everywhere.  They don't jump sharply when an interaction moves between two periodic copies.

Bonded interactions by default use neither cutoffs nor periodic boundary conditions.  There's an option to enable periodic boundary conditions.  That option *does not enable cutoffs*.  That means the forces do jump sharply when moving between two periodic copies.  It's intended to be used in situations where you're sure that will never happen: bonding one side of a molecule to the opposite side to create an infinite chain.

Which behavior do you want here?  Do you want to make the exceptions behave like nonbonded interactions, including smooth cutoffs?  Or do you want them to behave like bonded interactions with no cutoff, but with periodic boundary conditions enabled?

**Comment by kcui7** [comment]
I want the exceptions behave like the bonded interaction. Let me rephrase the problem. 

Suppose we have a infinite atom chain (or a 3D material), as shown in the scheme. The atom A is bonded to atom B across the boundary. Let B* denote the periodic image of atom B. 
![scheme1](https://user-images.githubusercontent.com/39925154/75054753-4e620e80-5499-11ea-84fb-1c0acb008d6c.png)

The right way to calculate the interaction between A and B (and it's images) should be as follows:
-  Calculate the bonded interaction between A and B*;
-  Calculate the exception between A and B*;
-  Calculate the full nonbonded interaction between A and B. 

However, in OpenMM these interactions are implemented in the following way:
-  Calculate the bonded interaction between A and B*; (correct)
-  Calculate the exception between A and B; (incorrect)
-  Calculate the full nonbonded interaction between A and B*. (incorrect)

I think the problem is that when define exceptions of one atom, OpenMM only records the index of the other atoms involved in the exception, but not record in which image the other atoms should be in.

**Comment by peastman** [comment]
Ok, I'll make it work the same way bonded interactions do.  Thanks for the quick response.

**Comment by peastman** [comment]
I just submitted the PR: #2570.
